### PR TITLE
Allow for custom packages data providers

### PR DESCRIPTION
**Pull request by marmichalski**

### CODE REVIEW

File: src/ComposerRunner/SymfonyProcessComposerRunner.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+declare(strict_types=1);
+
+namespace Lendable\ComposerLicenseChecker\ComposerRunner;
+
+use Lendable\ComposerLicenseChecker\ComposerRunner;
+use Lendable\ComposerLicenseChecker\Exception\FailedRunningComposer;
+use Symfony\Component\Process\Process;
+
+final class SymfonyProcessComposerRunner implements ComposerRunner
```
Comment (kunicmarko20) [inline src/ComposerRunner/SymfonyProcessComposerRunner.php:11]: why is this "Symfony specific" ?

File: src/Exception/FailedRunningComposer.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Lendable\ComposerLicenseChecker\Exception;
+
+final class FailedRunningComposer extends \RuntimeException
+{
+    private function __construct(string $message, ?\Throwable $previous = null)
```
Comment (ben-challis) [inline src/Exception/FailedRunningComposer.php:9]:
```suggestion
    /**
     * @param non-empty-string $message
     */
    private function __construct(string $message, ?\Throwable $previous = null)
```

File: src/Exception/FailedRunningComposer.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        parent::__construct($message, 0, $previous);
+    }
+
+    public static function withReason(string $reason): self
```
Comment (ben-challis) [inline src/Exception/FailedRunningComposer.php:14]:
```suggestion
    /**
     * @param non-empty-string $reason
     */
    public static function withReason(string $reason): self
```

File: src/Exception/InvalidPackageName.php [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+namespace Lendable\ComposerLicenseChecker\Exception;
+
+final class InvalidPackageName extends \RuntimeException
+{
+    private function __construct(string $message, ?\Throwable $previous = null)
+    {
+        parent::__construct($message, 0, $previous);
+    }
+
+    public static function for(string $packageName): self
```
Comment (ben-challis) [inline src/Exception/InvalidPackageName.php:14]:
```suggestion
    /**
     * @param non-empty-string $packageName
     */
    public static function for(string $packageName): self
```

File: src/Exception/PackagesProviderNotLocated.php [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Lendable\ComposerLicenseChecker\Exception;
+
+final class PackagesProviderNotLocated extends \RuntimeException
+{
+    private function __construct(string $message, ?\Throwable $previous = null)
```
Comment (ben-challis) [inline src/Exception/PackagesProviderNotLocated.php:9]:
```suggestion
    /**
     * @param non-empty-string $message
     */
    private function __construct(string $message, ?\Throwable $previous = null)
```

File: src/Exception/PackagesProviderNotLocated.php [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        parent::__construct($message, 0, $previous);
+    }
+
+    public static function withId(string $id): self
```
Comment (ben-challis) [inline src/Exception/PackagesProviderNotLocated.php:14]:
```suggestion
    /**
     * @param non-empty-string $id
     */
    public static function withId(string $id): self
```

File: src/InMemoryPackagesProviderLocator.php [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+{
+    /**
+     * @param array<string, PackagesProvider> $providers
+     */
+    public function __construct(private readonly array $providers)
+    {
+    }
+
+    /**
+     * @return list<string>
```
Comment (ben-challis) [inline src/InMemoryPackagesProviderLocator.php:19]:
```suggestion
     * @return list<non-empty-string>
```

File: src/InMemoryPackagesProviderLocator.php [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+declare(strict_types=1);
+
+namespace Lendable\ComposerLicenseChecker;
+
+use Lendable\ComposerLicenseChecker\Exception\PackagesProviderNotLocated;
+
+final class InMemoryPackagesProviderLocator implements PackagesProviderLocator
+{
+    /**
+     * @param array<string, PackagesProvider> $providers
```
Comment (ben-challis) [inline src/InMemoryPackagesProviderLocator.php:12]:
```suggestion
     * @param array<non-empty-string, PackagesProvider> $providers
```

File: src/LicenseChecker.php [hunk 9]
```diff
@@ -65,43 +76,42 @@ protected function execute(InputInterface $input, OutputInterface $output): int
             return self::FAILURE;
         }
 
-        $process = Process::fromShellCommandline('SHELL_VERBOSITY=0 composer licenses --format=json', $path);
-        $process->run();
-        if (!$process->isSuccessful()) {
-            $style->error(\sprintf('Failed to run "composer licenses --format=json" (%d).', $process->getExitCode()));
+        $path = (string) \realpath($path ?? \dirname(__DIR__));
+        /** @var string $providerId */
```
Comment (ben-challis) [inline src/LicenseChecker.php:80]:
```suggestion
        /** @var non-empty-string $providerId */
```

Though I think we should just check this at runtime fully instead.

File: src/Package.php [hunk 10]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Lendable\ComposerLicenseChecker;
+
+final class Package
+{
+    /**
+     * @param list<string> $licenses
```
Comment (ben-challis) [inline src/Package.php:10]:
```suggestion
     * @param list<non-empty-string> $licenses
```

File: src/Packages.php [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+{
+    /**
+     * @param array<Package> $items
+     */
+    public function __construct(
+        private readonly array $items,
+    ) {
+    }
+
+    public function sorted(): self
```
Comment (kunicmarko20) [inline src/Packages.php:20]: sorted vs sort?

File: src/PackagesProvider/ComposerInstalledJsonPackagesProvider.php [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+use Lendable\ComposerLicenseChecker\Exception\FailedProvidingPackages;
+use Lendable\ComposerLicenseChecker\Package;
+use Lendable\ComposerLicenseChecker\PackageName;
+use Lendable\ComposerLicenseChecker\Packages;
+use Lendable\ComposerLicenseChecker\PackagesProvider;
+
+final class ComposerInstalledJsonPackagesProvider implements PackagesProvider
+{
+    private const DS = \DIRECTORY_SEPARATOR;
```
Comment (kunicmarko20) [inline src/PackagesProvider/ComposerInstalledJsonPackagesProvider.php:15]: why not just use `\DIRECTORY_SEPARATOR`??

**Review by kunicmarko20 — APPROVED** [review]

File: src/PackagesProvider/ComposerLicensesPackagesProvider.php [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        } catch (FailedRunningComposer $e) {
+            throw FailedProvidingPackages::dueTo($e);
+        }
+
+        try {
+            /** @var array{
+             *      name: string,
+             *      version: string,
+             *      license: list<string>,
+             *      dependencies: array<string, array{version: string, license: list<string>}>
```
Comment (ben-challis) [inline src/PackagesProvider/ComposerLicensesPackagesProvider.php:34]:
```suggestion
             *      name: non-empty-string,
             *      version: non-empty-string,
             *      license: list<non-empty-string>,
             *      dependencies: array<non-empty-string, array{version: non-empty-string, license: list<non-empty-string>}>
```

File: src/PackagesProviderLocator.php [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+declare(strict_types=1);
+
+namespace Lendable\ComposerLicenseChecker;
+
+use Lendable\ComposerLicenseChecker\Exception\PackagesProviderNotLocated;
+
+interface PackagesProviderLocator
+{
+    /**
+     * @return list<string>
```
Comment (ben-challis) [inline src/PackagesProviderLocator.php:12]:
```suggestion
     * @return list<non-empty-string>
```

File: src/PackagesProviderLocator.php [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+     */
+    public function ids(): array;
+
+    /**
+     * @throws PackagesProviderNotLocated
+     */
+    public function locate(string $id): PackagesProvider;
```
Comment (ben-challis) [inline src/PackagesProviderLocator.php:19]:
```suggestion
    /**
     * @param non-empty-string $id
     */
    public function locate(string $id): PackagesProvider;
```

File: src/LicenseChecker.php (hunk 9 above)
Comment (marmichalski) [reply src/LicenseChecker.php:80]: This will fail while trying to locate, guess that's fine?

File: src/Package.php (hunk 10 above)
Comment (marmichalski) [reply src/Package.php:10]: `license` is optional though :(

**Pull request merged**